### PR TITLE
Improve doc cfg(test) and tests directory

### PR DESCRIPTION
**Pull request by Freyskeyd**
Hi,

I was facing a problem with my code organisation. I was using a tests directory and i defined some `#[cfg(test)]` in my `src/`. But i was not able to use it in my `tests` folder.

```bash
.
├── Cargo.lock
├── Cargo.toml
├── src
│   ├── lib.rs
│   └── test.rs
└── tests
    └── x.rs
```
> src/lib.rs
```rust
#[cfg(test)]
pub mod test;

#[test]
fn tesst() {
    assert!(test::t());
}
```
> src/test.rs
```rust
pub fn t() -> bool { true }
```
> test/x.rs
```rust
extern crate testt;

use testt::test;
#[test]
fn tesst() {
    assert!(test::t());
}
```

I was unable to compile using `cargo test`:
```bash
error[E0432]: unresolved import `testt::test`
 --> tests/x.rs:3:5
  |
3 | use testt::test;
  |     ^^^^^^^^^^^ no `test` in `testt`
```

If i remove the `tests` directory everything works fine. To use an utils module in your `tests` directory, you need to create a module in the directory (like `tests/utils.rs`). My `tests/x.rs` look like this now:

```rust
extern crate testt;

mod utils;

#[test]
fn tesst() {
    assert!(utils::t());
}
```

And my tree:
```bash
.
├── Cargo.lock
├── Cargo.toml
├── src
│   └── lib.rs
└── tests
    ├── utils.rs
    └── x.rs
```

I think that thing must be documented in the book.

Ping:
- @badboy : Because he's the one who showed me the path
- @shahn : Because he helped me too to find the solution

Signed-off-by: Freyskeyd <simon.paitrault@iadvize.com>

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I don't like very much the way you worded it. :-/ `#[cfg(test)]` isn't *ignored*, this is conditional compilation so it cannot be ignored.

**Comment by Freyskeyd** [comment]
@GuillaumeGomez right.

What do you think:

```
Note, every `#[cfg(test)]` in the `src` directory will not be compiled because...
```

**Comment by badboy** [comment]
The section already mentions: 
> This is because each test in the `tests` directory is an entirely separate crate, and so we need to import our library.

Maybe we should just stress a bit more that this means that the main crate is not compiled in test mode?

**Comment by Freyskeyd** [comment]
Something like: 

```
This is because each test in the `tests` directory is an entirely separate crate, and so we need to import our library.
Meaning that every `#[cfg (test)]` will not be compiled in the main crate.
```

cc @badboy

**Comment by GuillaumeGomez** [comment]
Hum, I'm shared on this. :-/

cc @rust-lang/docs

**Comment by frewsxcv** [comment]
I'm not a fan of the proposed wording for this, but I can't think of anything better to suggest right now.

**Comment by steveklabnik** [comment]
Here's how I'd word this:

> +Note, every `#[cfg(test)]` in the `src` directory will be ignored because we use
+separate crates in `tests` directory as we want to test the library like any
+other consumer.

> Note, when building integration tests, your library is compiled *without* `cfg(test)`, since it's an external test, and so is testing your library as consumers of that library would use it.

WDTY @rust-lang/docs ?

**Comment by GuillaumeGomez** [comment]
@steveklabnik: I don't like your wording either. Too confusing.

**Comment by steveklabnik** [comment]
Could you elaborate on what specifically is confusing?

**Comment by GuillaumeGomez** [comment]
Sure!

> your library is compiled without `cfg(test)`

I think a better wording would be:

> since the `test` flag isn't passed to the compiler, all parts in `cfg(test)` won't be included in the build

**Comment by steveklabnik** [comment]
I like that too. @Freyskeyd what do you think?

**Comment by Freyskeyd** [comment]
That's nice, what do you think of:

> Note, when a `tests` folder exists, cargo will not pass the `test`flag to the compiler, all parts in `cfg(test)` won't be included in the build.

**Comment by frewsxcv** [comment]
> Note, when a `tests` folder exists, cargo will not pass the `test`flag to the compiler, all parts in `cfg(test)` won't be included in the build.

@Freyskeyd The underlying issue here is not that a certain directory exists, but that when running integration tests, the library crate being tested will not be compiled with `cfg(test)`, which some users might be confused by.

---------

Regarding @steveklabnik's suggestion:

> Note, when building integration tests, your library is compiled *without* `cfg(test)`...
I like this part.

I like this part.

> ...since it's an external test, and so is testing your library as consumers of that library would use it.

I like the direction of this: indicating the library is being built in the same manner as a consumer might build it, but I'm not much a fan of the wording. I've read it through a dozen times and still don't really get it. I might be too sleepy right now though 😴 .

This is clearer to me, albeit maybe a bit wordy:

> Note, when building integration tests, your library is compiled *without* `cfg(test)`. Building your library without the `test` attribute better emulates how a consumer of your library might build it.

**Comment by GuillaumeGomez** [comment]
@frewsxcv: your sentence has the same issue:

> your library is compiled without `cfg(test)`

This is how it works, but it'd be great to include *why*, like I did when I proposed a modified version of what @steveklabnik suggested.

**Comment by Freyskeyd** [comment]
@GuillaumeGomez what do you think:

> Note, when building integration tests, cargo will not pass the `test` flag to the compiler. It means that all parts in cfg(test) won't be included in the build used in your integration tests.

**Comment by GuillaumeGomez** [comment]
@Freyskeyd: Great! 👍

File: src/doc/book/testing.md [hunk 1]
```diff
@@ -499,6 +499,10 @@ be imported in every test with `mod common;`
 That's all there is to the `tests` directory. The `tests` module isn't needed
 here, since the whole thing is focused on tests.
 
+Note, when building integration tests, cargo will not pass the test flag to the
+compiler. It means that all parts in cfg(test) won't be included in the build
```
Comment (frewsxcv) [inline src/doc/book/testing.md]: Nit:

```
`cfg(test)`
```

File: src/doc/book/testing.md [hunk 2]
```diff
@@ -499,6 +499,10 @@ be imported in every test with `mod common;`
 That's all there is to the `tests` directory. The `tests` module isn't needed
 here, since the whole thing is focused on tests.
 
+Note, when building integration tests, cargo will not pass the test flag to the
```
Comment (frewsxcv) [inline src/doc/book/testing.md]: Nit:

```
cargo will not pass the `test` attribute
```

https://doc.rust-lang.org/book/testing.html#the-test-attribute

Comment (Freyskeyd) [inline src/doc/book/testing.md]: @frewsxcv you want me to add a link to this page or add single quote around `test`?

Comment (frewsxcv) [inline src/doc/book/testing.md]: Adding the grave accents and calling it an "attribute" instead of a "flag", since TRPL calls it an "attribute".

Comment (frewsxcv) [inline src/doc/book/testing.md]: The reference also calls them "attributes":

https://doc.rust-lang.org/reference.html#function-only-attributes

Comment (Freyskeyd) [inline src/doc/book/testing.md]: yes, but in [conditional-compilation](https://doc.rust-lang.org/reference.html#conditional-compilation) it's a flag:

`test - Enabled when compiling the test harness (using the --test flag).`

Comment (frewsxcv) [inline src/doc/book/testing.md]: That's referring to `--test`, which is a command line argument, also called a "flag"

https://en.wikipedia.org/wiki/Command-line_interface#Command-line_option

**Comment by Freyskeyd** [comment]
@rust-lang/docs What do you think ? :)

**Comment by steveklabnik** [comment]
Looks fine to me, let's :shipit: 

@bors: r+ rollup
